akka-cluster-tools/src/multi-jvm/scala/akka/cluster/pubsub/DistributedPubSubRestartSpec.scala

```diff
@@ -0,0 +1,162 @@
+/**
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.cluster.pubsub
+
+import language.postfixOps
+import scala.concurrent.duration._
+import com.typesafe.config.ConfigFactory
+import akka.actor.Actor
+import akka.actor.ActorRef
+import akka.actor.PoisonPill
+import akka.actor.Props
+import akka.cluster.Cluster
+import akka.cluster.ClusterEvent._
+import akka.remote.testconductor.RoleName
+import akka.remote.testkit.MultiNodeConfig
+import akka.remote.testkit.MultiNodeSpec
+import akka.remote.testkit.STMultiNodeSpec
+import akka.testkit._
+import akka.actor.ActorLogging
+import akka.cluster.pubsub.DistributedPubSubMediator.Internal.Status
+import akka.cluster.pubsub.DistributedPubSubMediator.Internal.Delta
+import akka.actor.ActorSystem
+import scala.concurrent.Await
+import akka.actor.Identify
+import akka.actor.RootActorPath
+import akka.actor.ActorIdentity
+
+object DistributedPubSubRestartSpec extends MultiNodeConfig {
+  val first = role("first")
+  val second = role("second")
+  val third = role("third")
+
+  commonConfig(ConfigFactory.parseString("""
+    akka.loglevel = INFO
+    akka.cluster.pub-sub.gossip-interval = 500ms
+    akka.actor.provider = "akka.cluster.ClusterActorRefProvider"
+    akka.remote.log-remote-lifecycle-events = off
+    akka.cluster.auto-down-unreachable-after = off
+    """))
+
+  testTransport(on = true)
+
+  class Shutdown extends Actor {
+    def receive = {
+      case "shutdown" ⇒ context.system.terminate()
+    }
+  }
+
+}
+
+class DistributedPubSubRestartMultiJvmNode1 extends DistributedPubSubRestartSpec
+class DistributedPubSubRestartMultiJvmNode2 extends DistributedPubSubRestartSpec
+class DistributedPubSubRestartMultiJvmNode3 extends DistributedPubSubRestartSpec
+
+class DistributedPubSubRestartSpec extends MultiNodeSpec(DistributedPubSubRestartSpec) with STMultiNodeSpec with ImplicitSender {
+  import DistributedPubSubRestartSpec._
+  import DistributedPubSubMediator._
+
+  override def initialParticipants = roles.size
+
+  def join(from: RoleName, to: RoleName): Unit = {
+    runOn(from) {
+      Cluster(system) join node(to).address
+      createMediator()
+    }
+    enterBarrier(from.name + "-joined")
+  }
+
+  def createMediator(): ActorRef = DistributedPubSub(system).mediator
+  def mediator: ActorRef = DistributedPubSub(system).mediator
+
+  def awaitCount(expected: Int): Unit = {
+    val probe = TestProbe()
+    awaitAssert {
+      mediator.tell(Count, probe.ref)
+      probe.expectMsgType[Int] should ===(expected)
+    }
+  }
+
+  "A Cluster with DistributedPubSub" must {
+
+    "startup 3 node cluster" in within(15 seconds) {
+      join(first, first)
+      join(second, first)
+      join(third, first)
+      enterBarrier("after-1")
+    }
+
+    "handle restart of nodes with same address" in within(30 seconds) {
```

Comment:
Not easy to test. I had to be somewhat "creative" to write a test for the issue. This was failing before the fix.